src/mlpack/tests/main_tests/hmm_loglik_test.cpp

```diff
@@ -0,0 +1,165 @@
+    // Therefore we have to use the helper functions.
+    RandomInitialize(hmm.Emission());
+  }
+
+  //! Helper function to create discrete HMM.
+  static void Create(HMM<DiscreteDistribution>& hmm,
+                     vector<mat>& trainSeq,
+                     size_t states,
+                     double tolerance = 1e-05)
+  {
+    // Maximum observation is necessary so we know how to train the discrete
+    // distribution.
+    arma::Col<size_t> maxEmissions(trainSeq[0].n_rows);
+    maxEmissions.zeros();
+    for (vector<mat>::iterator it = trainSeq.begin(); it != trainSeq.end();
+         ++it)
+    {
+      arma::Col<size_t> maxSeqs =
+          arma::conv_to<arma::Col<size_t>>::from(arma::max(*it, 1)) + 1;
+      maxEmissions = arma::max(maxEmissions, maxSeqs);
+    }
+
+    hmm = HMM<DiscreteDistribution>(size_t(states),
+        DiscreteDistribution(maxEmissions), tolerance);
+  }
+
+  static void Create(HMM<GaussianDistribution>& hmm,
+                     vector<mat>& trainSeq,
+                     size_t states,
+                     double tolerance = 1e-05)
+  {
+    // Not implemented
+    // Prevent unused args warning
+    (void)hmm;
+    (void)trainSeq;
+    (void)states;
+    (void)tolerance;
+  }
+
+  static void Create(HMM<GMM>& hmm,
+                     vector<mat>& trainSeq,
+                     size_t states,
+                     double tolerance = 1e-05)
+  {
+    // Not implemented
+    // Prevent unused args warning
+    (void)hmm;
+    (void)trainSeq;
+    (void)states;
+    (void)tolerance;
+  }
+
+  //! Helper function for discrete emission distributions.
+  static void RandomInitialize(vector<DiscreteDistribution>& e)
+  {
+    for (size_t i = 0; i < e.size(); ++i)
+    {
+      e[i].Probabilities().randu();
+      e[i].Probabilities() /= arma::accu(e[i].Probabilities());
+    }
+  }
+
+  static void RandomInitialize(vector<GaussianDistribution>& e)
+  {
+    // Not implemented
+    // Prevent unused args warning
+    (void)e;
+  }
+
+  static void RandomInitialize(vector<GMM>& e)
+  {
+    // Not implemented
+    // Prevent unused args warning
+    (void)e;
+  }
+};
+
+struct Train
+{
+  template<typename HMMType>
+  static void Apply(HMMType& hmm, vector<arma::mat>* trainSeq)
+  {
+    // For now, perform unsupervised (Baum-Welch) training
+    hmm.Train(*trainSeq);
+  }
+};
+
+BOOST_AUTO_TEST_CASE(HMMLoglikOutputNegativeTest)
+{
+  // Create an HMMModel
+  HMMModel * h = new HMMModel(DiscreteHMM);
+  // Load data
+  arma::mat inp;
+  data::Load("obs1.csv", inp);
+  std::vector<arma::mat> trainSeq = {inp};
+  // Init HMMModel
+  h->PerformAction<Init, std::vector<arma::mat>>(&trainSeq);
+  // Train HMMModel
+  h->PerformAction<Train, std::vector<arma::mat>>(&trainSeq);
+
+
+  // Set the params for the hmm_loglik invocation
+  SetInputParam("input_model", h);
+  SetInputParam("input", inp);
+
+  mlpackMain();
+
+  double loglik = CLI::GetParam<double>("log_likelihood");
+  BOOST_REQUIRE(loglik <= 0);
```

Comment:
Are you sure it's guaranteed that the log-likelihood be less than 0?  I think any value that's not NaN or Inf here is fine.

Also, I think it might be worthwhile to check here also with `GaussianHMM` and `GaussianMixtureModelHMM`.

Reply:
Erm.. Isn't the product of all scaling factors equal to the probability of a given observation sequence ([ref](https://en.wikipedia.org/wiki/Forward%E2%80%93backward_algorithm#Forward_probabilities))? And the log of a probability should be <= 0. I'm fairly certain <= 0 is correct -- let me know what you think.

Reply:
I feel like I should elaborate, just so I'm conceptually clear on what is happening in the algorithm. In
https://github.com/mlpack/mlpack/blob/486be65dab54e8f0fba4066827c8358ffdea3d4e/src/mlpack/methods/hmm/hmm_impl.hpp#L454-L464
The call to `Forward()` populates `forward` and `scales`. Product of all elements in `scales` is:
 
<a href="https://www.codecogs.com/eqnedit.php?latex=P(o_1,o_2&space;...&space;o_k&space;|&space;\pi)" target="_blank"><img src="https://latex.codecogs.com/gif.latex?P(o_1,o_2&space;...&space;o_k&space;|&space;\pi)" title="P(o_1,o_2 ... o_k | \pi)" /></a>

And `LogLikelihood` returns the log of this probability.

Reply:
You're right, thanks for the clarification.  I think I wasn't thinking on all cylinders. :)